runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/state/WindmillStateInternalsTest.java

```diff
@@ -1030,6 +1075,64 @@ true, key(NAMESPACE, tag), STATE_FAMILY, VarIntCoder.of()))
     assertThat(keys, Matchers.containsInAnyOrder(key1, key2, key3));
   }
 
+  @Test
+  public void testMapViaMultimapEntriesAndKeysMergeLocalAddRemoveClear() {
+    final String tag = "map";
+    StateTag<MapState<byte[], Integer>> addr =
+        StateTags.map(tag, ByteArrayCoder.of(), VarIntCoder.of());
+    MapState<byte[], Integer> mapState = underTestMapViaMultimap.state(NAMESPACE, addr);
+
+    final byte[] key1 = "key1".getBytes(StandardCharsets.UTF_8);
+    final byte[] key2 = "key2".getBytes(StandardCharsets.UTF_8);
+    final byte[] key3 = "key3".getBytes(StandardCharsets.UTF_8);
+    final byte[] key4 = "key4".getBytes(StandardCharsets.UTF_8);
+
+    SettableFuture<Iterable<Map.Entry<ByteString, Iterable<Integer>>>> entriesFuture =
+        SettableFuture.create();
+    when(mockReader.multimapFetchAllFuture(
+            false, key(NAMESPACE, tag), STATE_FAMILY, VarIntCoder.of()))
+        .thenReturn(entriesFuture);
+    SettableFuture<Iterable<Map.Entry<ByteString, Iterable<Integer>>>> keysFuture =
+        SettableFuture.create();
+    when(mockReader.multimapFetchAllFuture(
+            true, key(NAMESPACE, tag), STATE_FAMILY, VarIntCoder.of()))
+        .thenReturn(keysFuture);
+
+    ReadableState<Iterable<Map.Entry<byte[], Integer>>> entriesResult =
+        mapState.entries().readLater();
+    ReadableState<Iterable<byte[]>> keysResult = mapState.keys().readLater();
+    waitAndSet(entriesFuture, Arrays.asList(multimapEntry(key1, 3), multimapEntry(key2, 4)), 30);
+    waitAndSet(keysFuture, Arrays.asList(multimapEntry(key1), multimapEntry(key2)), 30);
+
+    mapState.put(key1, 7);
+    mapState.put(dup(key3), 8);
+    mapState.put(key4, 1);
+    mapState.remove(key4);
```

Comment:
I would also be interesting to test the case where `key5` is returned by the backend read, but we've already called `mapState.remove(key5)` to make sure it's properly filtered out. I think this is basically the contents of `testMultimapEntriesAndKeysMergeLocalRemove` below, but not for MapViaMultimap.

Reply:
done